### ISSUE TITLE
Doing multiple Matrix Multiply using for loop vs doing a single Matrix Multiply concatenating the inputs, gives two different results

### ISSUE THREAD

**Issue by DiableJambe** [pytorch/pytorch#9978]
## Issue description

When performing matrix multiply on a 3d tensor, in two different ways, I get two different gradients when performing backward on the output . Please see code example below for exact problem statement.

Apologize if there is something wrong with my understanding, but kindly help me to the right resource for the correct understanding.

## Code example
```
import torch

x = torch.autograd.Variable(torch.FloatTensor(4,4,4));
x.data.normal_(0.0,0.1);

y = torch.autograd.Variable(torch.FloatTensor(4,5), requires_grad=True);
y.data.normal_(0.0,0.1);

x2 = torch.autograd.Variable(x.data);
z  = torch.autograd.Variable(y.data, requires_grad=True);

w1 = [];

for x_ in torch.unbind(x, dim=0):
    w1.append(torch.matmul(x, y));

w1 = torch.stack(w1);

w2 = torch.matmul(x2.view(16,4), z).view(4,4,5);

print(torch.sum((w2 - w1)**2));

loss1 = torch.sum((w1 - 5)**2);

loss2 = torch.sum((w2 - 5)**2);

loss1.backward();

loss2.backward();

print(y.grad);
print(z.grad);
```
## Output
```
Variable containing:
 0
[torch.FloatTensor of size (1,)]

Variable containing:
-10.6638 -10.6886 -10.9352 -10.5696 -10.7285
-19.2051 -19.1448 -19.1945 -19.3385 -19.2048
 -9.0791  -9.0517  -8.9411  -9.1330  -9.0585
 -4.8962  -4.7247  -4.6640  -5.0115  -4.8927
[torch.FloatTensor of size (4,5)]

Variable containing:
-2.6659 -2.6722 -2.7338 -2.6424 -2.6821
-4.8013 -4.7862 -4.7986 -4.8346 -4.8012
-2.2698 -2.2629 -2.2353 -2.2833 -2.2646
-1.2240 -1.1812 -1.1660 -1.2529 -1.2232
[torch.FloatTensor of size (4,5)]
```
## System Info

PyTorch version: 0.4.0a0+65fb885
Is debug build: No
CUDA used to build PyTorch: 8.0.44

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 7.5.17
GPU models and configuration:
GPU 0: TITAN Xp
GPU 1: TITAN Xp

Nvidia driver version: 390.25
cuDNN version: Probably one of the following:
/usr/local/lib64/libcudnn.so.5.1.5
/usr/local/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] numpy (1.14.0)
[pip] torch (0.4.0a0+65fb885)
[conda] magma-cuda80              2.2.0                h39f1f8d_1    pytorch
[conda] torch                     0.4.0a0+65fb885           <pip>

**Comment by DiableJambe** [comment]
Apologize. There was a bug in my code.
